### PR TITLE
New Actor: aws.iam.User

### PR DESCRIPTION
**Pull request by diranged**
Allows full management of Amazon IAM Users.

### CODE REVIEW

**Comment by siminm** [comment]
This breaks the original rule we set - not having any networking inside of the __init__

What's the advantage here? Tests aren't the only time we want to create those objects

**Comment by diranged** [comment]
So this is tricky -- if you're using things like temporary STS credentials, or you're using ~/.boto, or ~/.aws/credentials, our existing code didn't work. It explicitly required the AWS environment variables and completely bypassed the possibility of using boto's own credential logic.

The code still allows us to bypass the API calls in the init phase if we explicitly set the creds, but this change also allows you to properly leverage other boto credential patterns.

**Comment by siminm** [comment]
You'll create 4 real connections **per aws actor** on every run (dry/real). If you want to fool around on your dev you can populate the env variables expected by Kingpin. If you want to allow things like `.aws/credentials` credentials in kingpin then 1) make sure you add it to documentation, but also please figure out a way to make this lazy

**Comment by diranged** [comment]
@siminm  my read of the `boto.provider.Provider.__init__` method (https://github.com/boto/boto/blob/develop/boto/provider.py#L181-L214) was that as long as a user passes the AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY variables to the environment, that they would be loaded up first and the library would still be lazily instantiated. Only if they do _not_ pass these in (in the case of using `~/.aws.credentials`) would the `Provider` object make API calls. Am I misreading it?

File: kingpin/actors/aws/base.py [hunk 1]
```diff
@@ -83,19 +85,34 @@ def __init__(self, *args, **kwargs):
...
+        # By default, we will try to let Boto handle discovering its
+        # credentials at instantiation time. This _can_ result in synchronous
+        # API calls to the Metadata service, but those should be fast.
+        key = None
+        secret = None
+
+        # In the event though that someone has explicitly set the AWS access
+        # keys in the environment (either for the purposes of a unit test, or
+        # because they wanted to), we use those values.
+        if (aws_settings.AWS_ACCESS_KEY_ID and
```
Comment (siminm) [inline kingpin/actors/aws/base.py]: If you want to allow `boto` to do its thing then these variables are useless. Boto already checks environment variables by these names ( https://github.com/boto/boto/blob/develop/boto/provider.py#L282 )

File: examples/aws.iam.user/s3_example.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * easily in them!
+ */
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+
+            /***
+             * Note, these are out-of-order intetionally as part of our test
```
Comment (siminm) [inline examples/aws.iam.user/s3_example.json]: Out of what order?

File: kingpin/actors/aws/iam.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 from kingpin.actors.aws import base
 from kingpin.actors import exceptions
 from kingpin.constants import REQUIRED
+from kingpin.constants import STATE
```
Comment (siminm) [inline kingpin/actors/aws/iam.py]: Not sure how I like the kingpin-global STATE with 2 valid values. Wouldn't it be easier to just list the allowed values as the first entry in the all_options? Example:

```
'state': (('present','absent'), 'present', 'Desired state of the User: present/absent'),
```

Other actors may have other states, like 'public','private' etc

Comment (diranged) [inline kingpin/actors/aws/iam.py]: I have a thought on how to re-factor this so its more generic, and easier for other actors to use in the future. We can chat on Monday about it. Until then, ignore this code.

**Comment by diranged** [comment]
@siminm I have refactored the "iam.User" part of the code so that we don't throw unnecessary exceptions (which get logged in debug mode). Ignoring some of your other comments for now, you can at least take a look now at the pure IAM user creation part of the code.

## User doesn't exist, and we ensure they're absent
```
(.venv)Matts-MacBook-2:kingpin diranged$ PYTHONPATH=. python kingpin/bin/deploy.py --debug  --color  -a aws.iam.User -o name=mwise-test -o state=absent
11:36:50   INFO     71515 [__main__                                ] [main                ] Rehearsing... Break a leg!
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [DRY: Command Line] Checking for required options: ['name']
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [__init__            ] [DRY: Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Beginning
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [DRY: Command Line] Condition True evaluates to True
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [timeout             ] [DRY: Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:36:50   INFO     71515 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] Ensuring that user mwise-test is absent
11:36:50   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Searching for user mwise-test
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] User mwise-test not found
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] User mwise-test does not exist
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Finished successfully, return value: None
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [DRY: Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.87s
11:36:51   INFO     71515 [__main__                                ] [main                ] Rehearsal OK! Performing!
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [Command Line] Checking for required options: ['name']
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [__init__            ] [Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:36:51   INFO     71515 [__main__                                ] [main                ] 
11:36:51   WARNING  71515 [__main__                                ] [main                ] Lights, camera ... action!
11:36:51   INFO     71515 [__main__                                ] [main                ] 
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Beginning
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [Command Line] Condition True evaluates to True
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [timeout             ] [Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:36:51   INFO     71515 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] Ensuring that user mwise-test is absent
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Searching for user mwise-test
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] User mwise-test not found
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] User mwise-test does not exist
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Finished successfully, return value: None
11:36:51   DEBUG    71515 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.40s
```

## User doesn't exist, and we ensure is present
```
(.venv)Matts-MacBook-2:kingpin diranged$ PYTHONPATH=. python kingpin/bin/deploy.py --debug  --color  -a aws.iam.User -o name=mwise-test -o state=present
11:36:56   INFO     71518 [__main__                                ] [main                ] Rehearsing... Break a leg!
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [DRY: Command Line] Checking for required options: ['name']
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [__init__            ] [DRY: Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Beginning
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [DRY: Command Line] Condition True evaluates to True
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [timeout             ] [DRY: Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:36:56   INFO     71518 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] Ensuring that user mwise-test is present
11:36:56   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Searching for user mwise-test
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] User mwise-test not found
11:36:57   WARNING  71518 [kingpin.actors.aws.iam.User             ] [_create_user        ] [DRY: Command Line] Would create user mwise-test
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Finished successfully, return value: None
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [DRY: Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.41s
11:36:57   INFO     71518 [__main__                                ] [main                ] Rehearsal OK! Performing!
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [Command Line] Checking for required options: ['name']
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [__init__            ] [Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:36:57   INFO     71518 [__main__                                ] [main                ] 
11:36:57   WARNING  71518 [__main__                                ] [main                ] Lights, camera ... action!
11:36:57   INFO     71518 [__main__                                ] [main                ] 
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Beginning
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [Command Line] Condition True evaluates to True
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [timeout             ] [Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:36:57   INFO     71518 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] Ensuring that user mwise-test is present
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Searching for user mwise-test
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] User mwise-test not found
11:36:57   INFO     71518 [kingpin.actors.aws.iam.User             ] [_create_user        ] [Command Line] User arn:aws:iam::123123:user/mwise-test created
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Finished successfully, return value: None
11:36:57   DEBUG    71518 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.57s
```

## User exists and we ensure is present
```
(.venv)Matts-MacBook-2:kingpin diranged$ PYTHONPATH=. python kingpin/bin/deploy.py --debug  --color  -a aws.iam.User -o name=mwise-test -o state=present
11:37:14   INFO     71522 [__main__                                ] [main                ] Rehearsing... Break a leg!
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [DRY: Command Line] Checking for required options: ['name']
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [__init__            ] [DRY: Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Beginning
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [DRY: Command Line] Condition True evaluates to True
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [timeout             ] [DRY: Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:37:14   INFO     71522 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] Ensuring that user mwise-test is present
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Searching for user mwise-test
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Found user arn:aws:iam::123123:user/mwise-test
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] User mwise-test exists
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Finished successfully, return value: None
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [DRY: Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.42s
11:37:14   INFO     71522 [__main__                                ] [main                ] Rehearsal OK! Performing!
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [Command Line] Checking for required options: ['name']
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [__init__            ] [Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:37:14   INFO     71522 [__main__                                ] [main                ] 
11:37:14   WARNING  71522 [__main__                                ] [main                ] Lights, camera ... action!
11:37:14   INFO     71522 [__main__                                ] [main                ] 
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Beginning
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [Command Line] Condition True evaluates to True
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [timeout             ] [Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:37:14   INFO     71522 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] Ensuring that user mwise-test is present
11:37:14   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Searching for user mwise-test
11:37:15   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Found user arn:aws:iam::123123:user/mwise-test
11:37:15   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] User mwise-test exists
11:37:15   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Finished successfully, return value: None
11:37:15   DEBUG    71522 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.42s
```

## User exists, and we ensure is absent
```
(.venv)Matts-MacBook-2:kingpin diranged$ PYTHONPATH=. python kingpin/bin/deploy.py --debug  --color  -a aws.iam.User -o name=mwise-test -o state=absent
11:42:04   INFO     71556 [__main__                                ] [main                ] Rehearsing... Break a leg!
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [DRY: Command Line] Checking for required options: ['name']
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [__init__            ] [DRY: Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Beginning
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [DRY: Command Line] Condition True evaluates to True
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [timeout             ] [DRY: Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:42:04   INFO     71556 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [DRY: Command Line] Ensuring that user mwise-test is absent
11:42:04   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Searching for user mwise-test
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_get_user           ] [DRY: Command Line] Found user arn:aws:iam::123123:user/mwise-test
11:42:05   WARNING  71556 [kingpin.actors.aws.iam.User             ] [_delete_user        ] [DRY: Command Line] Would delete user mwise-test
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [execute             ] [DRY: Command Line] Finished successfully, return value: None
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [DRY: Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.92s
11:42:05   INFO     71556 [__main__                                ] [main                ] Rehearsal OK! Performing!
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_validate_options   ] [Command Line] Checking for required options: ['name']
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [__init__            ] [Command Line] Initialized (warn_on_failure=False, strict_init_context=True)
11:42:05   INFO     71556 [__main__                                ] [main                ] 
11:42:05   WARNING  71556 [__main__                                ] [main                ] Lights, camera ... action!
11:42:05   INFO     71556 [__main__                                ] [main                ] 
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Beginning
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_check_condition    ] [Command Line] Condition True evaluates to True
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [timeout             ] [Command Line] kingpin.actors.aws.iam.User._execute() deadline: 3600(s)
11:42:05   INFO     71556 [kingpin.actors.aws.iam.User             ] [_ensure_user        ] [Command Line] Ensuring that user mwise-test is absent
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Searching for user mwise-test
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_get_user           ] [Command Line] Found user arn:aws:iam::123123:user/mwise-test
11:42:05   INFO     71556 [kingpin.actors.aws.iam.User             ] [_delete_user        ] [Command Line] User mwise-test deleted
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [execute             ] [Command Line] Finished successfully, return value: None
11:42:05   DEBUG    71556 [kingpin.actors.aws.iam.User             ] [_wrap_in_timer      ] [Command Line] kingpin.actors.aws.iam.User.execute() execution time: 0.69s
```

## Integration tests too
```
(.venv)Matts-MacBook-2:kingpin diranged$ nosetests --include=integration --attr=integration -vv kingpin/actors/aws/test/integration_iam_users.py
nose.config: INFO: Ignoring files matching ['^\\.', '^_', '^setup\\.py$']
nose.config: INFO: Including tests matching ['integration']
nose.plugins.cover: INFO: Coverage report will include only packages: ['kingpin']
integration_01_ensure_user_absent (kingpin.actors.aws.test.integration_iam_users.IntegrationIAMUsers) ... ok
integration_02a_create_user_dry (kingpin.actors.aws.test.integration_iam_users.IntegrationIAMUsers) ... ok
integration_02b_create_user (kingpin.actors.aws.test.integration_iam_users.IntegrationIAMUsers) ... ok
integration_09_ensure_user_absent (kingpin.actors.aws.test.integration_iam_users.IntegrationIAMUsers) ... ok
```

File: kingpin/actors/aws/base.py (hunk 1 above)
Comment (diranged) [inline kingpin/actors/aws/base.py]: The only reason I put in the option for overriding it was for the unit tests -- its easier to do the forcing of the settings in the aws.settings module and then reloading that, than actually mocking everything out. Maybe there is a more clever way to globally mock out the aws settings so that we know for sure that during a unit test we're not reaching out to an API.

File: kingpin/actors/aws/iam.py (hunk 3 above)
Comment (diranged) [inline kingpin/actors/aws/iam.py]: I took a stab at refactoring this so its more powerful down the road.. the new mechanism will make it easy for individual actors to create complex option validation methods without cluttering up their individual actor classes.

File: kingpin/actors/aws/base.py (hunk 1 above)
Comment (siminm) [inline kingpin/actors/aws/base.py]: That makes sense, but if no variables are passed in then maybe tests can
populate the expected environment variables? Not sure about the best
practice here

**Comment by diranged** [comment]
@siminm this is finally ready for its first full review.. all functionality exists now.

File: kingpin/actors/aws/base.py [hunk 4]
```diff
@@ -216,3 +234,58 @@ def _get_meta_data(self, key):
...
+            A dictionary of the parsed policy.
+        """
+        # Run through any supplied Inline IAM Policies and verify that they're
+        # not corrupt very early on.
+        self.log.debug('Parsing and validating %s' % policy)
+
+        try:
+            p_doc = utils.convert_json_to_dict(json_file=policy,
+                                               tokens=os.environ)
+        except (LookupError, ValueError, IOError) as e:
```
Comment (siminm) [inline kingpin/actors/aws/base.py]: Seems like this should be handled inside of the convert_json_to_dict

File: kingpin/actors/aws/iam/base.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+# This executor is used by the tornado.concurrent.run_on_executor()
+# decorator. We would like this to be a class variable so its shared
+# across RightScale objects, but we see testing IO errors when we
+# do this.
+EXECUTOR = concurrent.futures.ThreadPoolExecutor(10)
+
+
+class IAMBaseActor(base.AWSBaseActor):
```
Comment (siminm) [inline kingpin/actors/aws/iam/base.py]: If all code was yanked out of here - why is this needed?

File: kingpin/actors/aws/iam/entities.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+
+
+# This executor is used by the tornado.concurrent.run_on_executor()
+# decorator. We would like this to be a class variable so its shared
+# across RightScale objects, but we see testing IO errors when we
+# do this.
+EXECUTOR = concurrent.futures.ThreadPoolExecutor(10)
+
+
+class EntityBaseActor(base.IAMBaseActor):
```
Comment (siminm) [inline kingpin/actors/aws/iam/entities.py]: There's a ton of try/catch-boto cases here. I think it could be simplified if you had a wrapper around `self.thread` something like `self.thread_boto` which would try/catch boto specific issues, and maybe handle retry on common cases there.

File: kingpin/utils.py [hunk 7]
```diff
@@ -347,6 +347,30 @@ def convert_json_to_dict(json_file, tokens):
     return decoded
 
 
+def order_dict(obj):
```
Comment (siminm) [inline kingpin/utils.py]: So a bit of nit-picking but let's be correct here. Python dictionaries don't have a concept of an order. Here's an example:

```
In [1]: {'foo1': 'bar1', 'foo2': 'bar2'} == {'foo2': 'bar2', 'foo1': 'bar1'}
Out[1]: True
```

What does have order is lists. 

```
In [2]: ['a', 'b'] == ['b', 'a']
Out[2]: False
```

So the helper part here that's being leveraged is ordered lists (be they inside of other lists or other dicts). But saying Ordered Dict is just not correct

File: kingpin/actors/aws/base.py (hunk 1 above)
Comment (diranged) [inline kingpin/actors/aws/base.py]: Do you have any thoughts yet on this? I think you were thinking about how to mock out the environment during the tests?

File: examples/aws.iam.user/s3_example.json (hunk 2 above)
Comment (diranged) [inline examples/aws.iam.user/s3_example.json]: Addressed with an updated comment.

File: kingpin/actors/aws/base.py (hunk 4 above)
Comment (diranged) [inline kingpin/actors/aws/base.py]: Handled this in a new commit..

File: kingpin/actors/aws/iam/base.py (hunk 5 above)
Comment (diranged) [inline kingpin/actors/aws/iam/base.py]: This actually never had any content -- but it was always there as a base.

File: kingpin/actors/aws/iam/entities.py (hunk 6 above)
Comment (diranged) [inline kingpin/actors/aws/iam/entities.py]: Per our discussion -- I am goign to punt on this until we can come up with a good design pattern here

File: kingpin/utils.py (hunk 7 above)
Comment (siminm) [inline kingpin/utils.py]: More examples if you're not convinced:

```
In [4]: {'foo': ['bar', 'biz', 'baz']} == {'foo': ['bar', 'biz', 'baz']}
Out[4]: True

In [5]: {'foo': ['bar', 'biz', 'baz']} == {'foo': ['biz', 'baz', 'bar']}
Out[5]: False
```

Comment (diranged) [inline kingpin/utils.py]: Ok well either way -- would you like to rename this? I don't really have a strong opinion here, its just a quick helper method.

**Comment by diranged** [comment]
@siminm rebased this off `master` (with the new optional integration work code)..

**Comment by siminm** [comment]
👍
